benches/benches.rs

```diff
@@ -36,25 +39,95 @@ fn criterion_benchmark(c: &mut Criterion) {
     let logistic_map_function_id =
         find_function_id(&logistic_map, "logistic_map::logistic_map::main");
 
-    c.bench_function("Cached JIT factorial_2M", |b| {
-        b.iter(|| jit_factorial.invoke_dynamic(factorial_function_id, &[], Some(u128::MAX)));
-    });
-    c.bench_function("Cached JIT fib_2M", |b| {
-        b.iter(|| jit_fibonacci.invoke_dynamic(fibonacci_function_id, &[], Some(u128::MAX)));
-    });
-    c.bench_function("Cached JIT logistic_map", |b| {
-        b.iter(|| jit_logistic_map.invoke_dynamic(logistic_map_function_id, &[], Some(u128::MAX)));
-    });
-
-    c.bench_function("Cached AOT factorial_2M", |b| {
-        b.iter(|| aot_factorial.invoke_dynamic(factorial_function_id, &[], Some(u128::MAX)));
-    });
-    c.bench_function("Cached AOT fib_2M", |b| {
-        b.iter(|| aot_fibonacci.invoke_dynamic(fibonacci_function_id, &[], Some(u128::MAX)));
-    });
-    c.bench_function("Cached AOT logistic_map", |b| {
-        b.iter(|| aot_logistic_map.invoke_dynamic(logistic_map_function_id, &[], Some(u128::MAX)));
-    });
+    let factorial_runner = load_contract_for_vm("programs/benches/factorial_2M.cairo");
+    let fibonacci_runner = load_contract_for_vm("programs/benches/fib_2M.cairo");
+    let logistic_map_runner = load_contract_for_vm("programs/benches/logistic_map.cairo");
+
+    let factorial_function = factorial_runner
+        .find_function("main")
+        .expect("failed to find main factorial function");
+    let fibonacci_function = fibonacci_runner
+        .find_function("main")
+        .expect("failed to find main fibonacci function");
+    let logistic_map_function = logistic_map_runner
+        .find_function("main")
+        .expect("failed to find main logistic map function");
+
+    {
+        let mut factorial_group = c.benchmark_group("factorial_2M");
+
+        factorial_group.bench_function("Cached JIT", |b| {
+            b.iter(|| jit_factorial.invoke_dynamic(factorial_function_id, &[], Some(u128::MAX)));
+        });
+        factorial_group.bench_function("Cached AOT", |b| {
+            b.iter(|| aot_factorial.invoke_dynamic(factorial_function_id, &[], Some(u128::MAX)));
+        });
+
+        factorial_group.bench_function("VM", |b| {
+            b.iter(|| {
+                factorial_runner.run_function_with_starknet_context(
```

Comment:
We need to find another way to execute the VM.
See the `fn run_function_with_starknet_context`
```
        let assembled_program = self.casm_program.clone().assemble_ex(&entry_code, &footer);
```
It clone the casm_program adding lot of overhead to the execution

We need to discuss how we do this, maybe we have to add a new method to the `cairo-vm/cairo1-run` crate

Reply:
I think that particular issue has been fixed: https://github.com/starkware-libs/cairo/blob/226c3646be247828c08b737a73f6da10de76e0fc/crates/cairo-lang-runner/src/lib.rs#L191
```rust
let assembled_cairo_program = self.casm_program.assemble_ex(&header, &footer);
```